File: rust/linkedspec-core/tests/rule_local_cursor_normalization_test.rs
use linkedspec_core::ast::{BodyElementKind, RuleMode};
use linkedspec_core::compiler::compile;
use linkedspec_core::parser::parse_spec;
use linkedspec_core::types::CompiledSpec;
use linkedspec_core::validation::validate;
use serde_json::Value;
use std::collections::BTreeSet;

const CONTRACT_SOURCE: &str =
    include_str!("../../../capability_conformance/rule_local_cursor_contract.json");

fn contract() -> Value {
    serde_json::from_str(CONTRACT_SOURCE).expect("rule-local cursor contract must be valid JSON")
}

fn edge_source(parent_family: &str, sources: &[&str], declared_rules: &[&str]) -> String {
    let mut source = if parent_family == "and" {
        "Top::AND\n".to_string()
    } else {
        "Top::\n".to_string()
    };
    for edge in sources {
        source.push(' ');
        source.push_str(edge);
        source.push('\n');
    }
    for label in declared_rules {
        source.push('\n');
        source.push_str(label);
        source.push_str(":\n /x/ /y/\n");
    }
    source
}

fn diagnostic_for(source: &str) -> linkedspec_core::error::PortableDiagnostic {
    let parsed = parse_spec(source).expect("diagnostic fixture must parse to typed AST");
    validate(&parsed)
        .expect_err("diagnostic fixture must be rejected")
        .diagnostic()
        .expect("failure must carry a portable diagnostic")
        .clone()
}

#[test]
fn contract_family_cases_use_exact_authored_classification() {
    let contract = contract();
    for case in contract["family_cases"].as_array().unwrap() {
        let id = case["id"].as_str().unwrap();
        let source = format!("{}\n /x/\n", case["header"].as_str().unwrap());
        let parsed = parse_spec(&source).unwrap_or_else(|error| panic!("{id}: {error}"));
        let mode = &parsed.rules[0].header.mode;
        assert_eq!(mode.is_and(), case["family"] == "and", "{id}");
    }

    assert!(!RuleMode::Pipe.is_and(), "compact | is authored OR");
    assert!(RuleMode::Single.is_and(), "compact & is authored AND");
}

#[test]
fn contract_edge_cases_parse_validate_and_lower_exactly() {
    let contract = contract();
    let diagnostics = contract["diagnostics"].as_array().unwrap();

    for case in contract["edge_resolution_cases"].as_array().unwrap() {
        let id = case["id"].as_str().unwrap();
        let source_text = case["source"].as_str().unwrap();
        let declared = case["declared_rules"]
            .as_array()
            .unwrap()
            .iter()
            .map(|label| label.as_str().unwrap())
            .collect::<Vec<_>>();
        let source = edge_source(
            case["parent_family"].as_str().unwrap(),
            &[source_text],
            &declared,
        );

        if let Some(expected_code) = case.get("expected_error").and_then(Value::as_str) {
            let diagnostic = diagnostic_for(&source);
            let diagnostic_contract = diagnostics
                .iter()
                .find(|row| row["code"] == expected_code)
                .unwrap();
            assert_eq!(diagnostic.code, expected_code, "{id}");
            assert_eq!(
                diagnostic.stage,
                diagnostic_contract["stage"].as_str().unwrap(),
                "{id}"
            );
            let expected_fields = diagnostic_contract["fields"]
                .as_array()
                .unwrap()
                .iter()
                .map(|field| field.as_str().unwrap())
                .collect::<BTreeSet<_>>();
            let actual_fields = diagnostic
                .fields
                .keys()
                .map(String::as_str)
                .collect::<BTreeSet<_>>();
            assert_eq!(actual_fields, expected_fields, "{id}: diagnostic fields");

            assert_eq!(
                diagnostic.field("rule_label"),
                Some(&serde_json::json!("Top")),
                "{id}: rule attribution"
            );
            if expected_fields.contains("target") {
                let expected_target = if expected_code == "bare_edge_target_undefined" {
                    "Missing"
                } else {
                    "Child"
                };
                assert_eq!(
                    diagnostic.field("target"),
                    Some(&serde_json::json!(expected_target)),
                    "{id}: target attribution"
                );
            }
            if expected_fields.contains("regex_index") {
                assert_eq!(
                    diagnostic.field("regex_index"),
                    Some(&serde_json::json!(0)),
                    "{id}: explicit zero index must be retained"
                );
            }
            if expected_fields.contains("targets") {
                assert_eq!(
                    diagnostic.field("targets"),
                    Some(&serde_json::json!(declared)),
                    "{id}: grouped target order"
                );
            }

            let serialized = serde_json::to_string(&diagnostic).unwrap();
            let roundtrip: linkedspec_core::error::PortableDiagnostic =
                serde_json::from_str(&serialized).unwrap();
            assert_eq!(roundtrip, diagnostic, "{id}: diagnostic JSON roundtrip");
            continue;
        }

        let parsed = parse_spec(&source).unwrap_or_else(|error| panic!("{id}: {error}"));
        validate(&parsed).unwrap_or_else(|error| panic!("{id}: {error}"));
        let expected = &case["expected"];
        if expected["kind"] == "lifecycle" {
            assert!(matches!(
                parsed.rules[0].body[0].kind,
                BodyElementKind::LifecycleMarker { .. }
            ));
            continue;
        }

        if expected["source_form"] == "bare" {
            let BodyElementKind::BareEdge {
                targets,
                code,
                fluent_chain,
            } = &parsed.rules[0].body[0].kind
            else {
                panic!("{id}: bare source was not retained as typed BareEdge");
            };
            assert_eq!(targets.len(), expected["targets"].as_array().unwrap().len());
            assert_eq!(code.is_some(), expected["has_block"].as_bool().unwrap());
            let expected_fluent = expected["targets"][0]["fluent"].as_str();
            assert_eq!(
                fluent_chain.first().map(|call| {
                    if call.args.is_empty() {
                        call.method.clone()
                    } else {
                        format!("{}({})", call.method, call.args)
                    }
                }),
                expected_fluent.map(str::to_string),
                "{id}"
            );
            assert!(
                parsed.rules[0]
                    .body
                    .iter()
                    .all(|element| !matches!(element.kind, BodyElementKind::Raw { .. })),
                "{id}: governed bare source must never fall back to Raw"
            );
        }

        let compiled = compile(&parsed).unwrap_or_else(|error| panic!("{id}: {error}"));
        let serialized = serde_json::to_string(&compiled).unwrap();
        let compiled: CompiledSpec = serde_json::from_str(&serialized).unwrap();
        let top = &compiled.rules[0];
        let expected_targets = expected["targets"].as_array().unwrap();
        if expected["ownership"] == "action" {
            assert_eq!(top.acode_dispatch.len(), expected_targets.len(), "{id}");
            assert!(top.bcode_dispatch.is_empty(), "{id}");
            for (actual, expected) in top.acode_dispatch.iter().zip(expected_targets) {
                assert_eq!(actual.child_label, expected["label"], "{id}");
                assert_eq!(
                    actual.child_regex_idx,
                    expected["index"].as_u64().unwrap_or(0) as usize,
                    "{id}"
                );
            }
        } else {
            assert_eq!(top.bcode_dispatch.len(), expected_targets.len(), "{id}");
            assert!(top.acode_dispatch.is_empty(), "{id}");
            for (actual, expected) in top.bcode_dispatch.iter().zip(expected_targets) {
                assert_eq!(actual.child_label, expected["label"], "{id}");
            }
        }
    }
}

#[test]
fn contract_rule_edge_sets_validate_normalized_ownership() {
    let contract = contract();
    for case in contract["rule_edge_set_cases"].as_array().unwrap() {
        let id = case["id"].as_str().unwrap();
        let sources = case["sources"]
            .as_array()
            .unwrap()
            .iter()
            .map(|source| source.as_str().unwrap())
            .collect::<Vec<_>>();
        let declared = case["declared_rules"]
            .as_array()
            .unwrap()
            .iter()
            .map(|label| label.as_str().unwrap())
            .collect::<Vec<_>>();
        let source = edge_source(case["parent_family"].as_str().unwrap(), &sources, &declared);

        if let Some(expected_code) = case.get("expected_error").and_then(Value::as_str) {
            let diagnostic = diagnostic_for(&source);
            assert_eq!(diagnostic.code, expected_code, "{id}");
            assert_eq!(diagnostic.stage, "validate_rule", "{id}");
            assert_eq!(
                diagnostic
                    .fields
                    .keys()
                    .map(String::as_str)
                    .collect::<BTreeSet<_>>(),
                BTreeSet::from(["ownerships", "rule_label"]),
                "{id}: diagnostic fields"
            );
            assert_eq!(
                diagnostic.field("ownerships"),
                Some(&serde_json::json!(["action", "blind"])),
                "{id}"
            );
            continue;
        }

        let parsed = parse_spec(&source).unwrap_or_else(|error| panic!("{id}: {error}"));
        validate(&parsed).unwrap_or_else(|error| panic!("{id}: {error}"));
        let compiled = compile(&parsed).unwrap_or_else(|error| panic!("{id}: {error}"));
        let top = &compiled.rules[0];
        if case["expected_ownership"] == "action" {
            assert!(!top.acode_dispatch.is_empty(), "{id}");
            assert!(top.bcode_dispatch.is_empty(), "{id}");
        } else {
            assert!(!top.bcode_dispatch.is_empty(), "{id}");
            assert!(top.acode_dispatch.is_empty(), "{id}");
        }
    }
}

#[test]
fn complete_line_header_rest_and_multiline_bare_edges_remain_typed() {
    for source in [
        "Top::\n Child\n\nChild:\n /x/\n",
        "Top:: Child\n\nChild:\n /x/\n",
        "Top::AND\n Child {\n  return(child_result)\n }\n\nChild:\n /x/\n",
    ] {
        let parsed = parse_spec(source).unwrap();
        assert!(matches!(
            parsed.rules[0].body[0].kind,
            BodyElementKind::BareEdge { .. }
        ));
        assert!(
            parsed.rules[0]
                .body
                .iter()
                .all(|element| !matches!(element.kind, BodyElementKind::Raw { .. }))
        );
    }

    let same_line_suffix = parse_spec("Top:: /x/ Child\n\nChild:\n /x/\n").unwrap();
    assert!(
        same_line_suffix.rules[0]
            .body
            .iter()
            .all(|element| !matches!(element.kind, BodyElementKind::BareEdge { .. })),
        "bare recognition is physical-line scoped"
    );
}

#[test]
fn compiler_lowers_family_ownership_into_typed_dispatch_tables() {
    let default_spec = parse_spec("Top::\n Child\n\nChild:\n /x/\n").unwrap();
    validate(&default_spec).unwrap();
    let default_compiled = compile(&default_spec).unwrap();
    assert_eq!(default_compiled.rules[0].acode_dispatch.len(), 1);
    assert!(default_compiled.rules[0].bcode_dispatch.is_empty());

    let and_spec = parse_spec("Top::AND\n Child\n\nChild:\n /x/\n").unwrap();
    validate(&and_spec).unwrap();
    let and_compiled = compile(&and_spec).unwrap();
    assert_eq!(and_compiled.rules[0].bcode_dispatch.len(), 1);
    assert!(and_compiled.rules[0].acode_dispatch.is_empty());

    let explicit_blind = parse_spec("Top::\n => Child\n\nChild:\n /x/\n").unwrap();
    validate(&explicit_blind).unwrap();
    let explicit_blind = compile(&explicit_blind).unwrap();
    assert_eq!(explicit_blind.rules[0].bcode_dispatch.len(), 1);
}

File: lua/test/rule_local_cursor_normalization_test.lua
local ast = require("linkedspec.spec_ast")
local compiled_spec = require("linkedspec.compiled_spec")
local json = require("linkedspec.json")
local spec_parser = require("linkedspec.spec_parser")
local spec_validator = require("linkedspec.spec_validator")

local linkedspec = {
  spec_ast = ast,
  json = json,
  parse_spec = spec_parser.parse_spec,
  validate_spec = spec_validator.validate_spec,
  is_spec_validation_error = spec_validator.is_validation_error,
  spec_validation_error_to_json = spec_validator.validation_error_to_json,
  compile_spec = compiled_spec.compile_spec,
}

local assertions = 0
local failures = {}

local function check(condition, label)
  assertions = assertions + 1
  if not condition then failures[#failures + 1] = label or "assertion failed" end
end

local function check_equal(actual, expected, label)
  check(actual == expected, (label or "values differ") ..
    ": expected " .. tostring(expected) .. ", got " .. tostring(actual))
end

local function check_same_json(actual, expected, label)
  check_equal(json.encode(actual), json.encode(expected), label)
end

local function read_file(path)
  local handle = assert(io.open(path, "rb"))
  local value = assert(handle:read("*a"))
  assert(handle:close())
  return value
end

local function capture(operation)
  local ok, value = pcall(operation)
  return ok, value
end

local function array_strings(values)
  local result = {}
  for index, value in ipairs(values) do result[index] = value end
  return result
end

local function sorted_keys(object)
  local result = {}
  for key in pairs(object) do result[#result + 1] = key end
  table.sort(result)
  return table.concat(result, ",")
end

local function sorted_values(values)
  local result = array_strings(values)
  table.sort(result)
  return table.concat(result, ",")
end

local function edge_source(parent_family, sources, declared_rules)
  local lines = { parent_family == "and" and "Top::AND" or "Top::" }
  for _, source in ipairs(sources) do lines[#lines + 1] = " " .. source end
  for _, label in ipairs(declared_rules) do
    lines[#lines + 1] = ""
    lines[#lines + 1] = label .. ":"
    lines[#lines + 1] = " /x/ /y/"
  end
  return table.concat(lines, "\n") .. "\n"
end

local function validation_failure(source)
  local ok, failure = capture(function()
    return linkedspec.validate_spec(linkedspec.parse_spec(source))
  end)
  check_equal(ok, false, "expected validation failure")
  check_equal(linkedspec.is_spec_validation_error(failure), true, "portable validation type")
  return failure
end

local contract = json.decode(read_file("capability_conformance/rule_local_cursor_contract.json"))
local diagnostics = {}
for _, row in ipairs(contract.diagnostics) do diagnostics[row.code] = row end

check_equal(#contract.family_cases, 36, "family row count")
check_equal(#contract.edge_resolution_cases, 18, "edge row count")
check_equal(#contract.rule_edge_set_cases, 6, "ownership-set row count")

for _, row in ipairs(contract.family_cases) do
  local parsed = linkedspec.parse_spec(row.header .. "\n /x/\n")
  local expected_and = row.family == "and"
  check_equal(ast.rule_mode_is_and(parsed.rules[1].header.mode), expected_and, row.id .. " AST family")
  local compiled = linkedspec.compile_spec(parsed, { validate_source = false }):rule("Top")
  check_equal(compiled.mode_metadata.is_and, expected_and, row.id .. " compiled family")
end
check_equal(ast.rule_mode_is_and(ast.rule_mode("Pipe")), false, "compact pipe is OR/default")
check_equal(ast.rule_mode_is_and(ast.rule_mode("Single")), true, "compact ampersand is AND")

for _, row in ipairs(contract.edge_resolution_cases) do
  local source = edge_source(row.parent_family, { row.source }, array_strings(row.declared_rules))
  if row.expected_error ~= nil then
    local failure = validation_failure(source)
    local expected_diagnostic = diagnostics[row.expected_error]
    local projected = linkedspec.spec_validation_error_to_json(failure)
    check_equal(failure.code, row.expected_error, row.id .. " diagnostic code")
    check_equal(failure.stage, expected_diagnostic.stage, row.id .. " diagnostic stage")
    check_equal(sorted_keys(projected.fields), sorted_values(expected_diagnostic.fields), row.id .. " fields")
    check_equal(projected.fields.rule_label, "Top", row.id .. " rule label")
    if projected.fields.target ~= nil then
      check_equal(
        projected.fields.target,
        row.expected_error == "bare_edge_target_undefined" and "Missing" or "Child",
        row.id .. " target"
      )
    end
    if projected.fields.regex_index ~= nil then
      check_equal(projected.fields.regex_index, 0, row.id .. " regex index")
    end
    if projected.fields.targets ~= nil then
      check_same_json(projected.fields.targets, row.declared_rules, row.id .. " targets")
    end
    check_same_json(json.decode(json.encode(projected)), projected, row.id .. " diagnostic JSON roundtrip")
  else
    local parsed = linkedspec.parse_spec(source)
    local valid, validation_error = capture(function() return linkedspec.validate_spec(parsed) end)
    check_equal(valid, true, row.id .. " validates: " .. tostring(validation_error))
    local expected = row.expected
    if valid then
      local first_kind = parsed.rules[1].body[1].kind
      if expected.kind == "lifecycle" then
        check_equal(ast.node_type(first_kind), "LifecycleMarkerBodyElementKind", row.id .. " lifecycle kind")
        check_equal(first_kind.marker, expected.name, row.id .. " lifecycle name")
      else
        if expected.source_form == "bare" then
          check_equal(ast.node_type(first_kind), "BareEdgeBodyElementKind", row.id .. " typed bare kind")
          check_equal(#first_kind.targets, #expected.targets, row.id .. " bare target count")
          check_equal(first_kind.code ~= nil, expected.has_block, row.id .. " bare block")
          local expected_fluent = expected.targets[1].fluent
          local actual_fluent = nil
          if #first_kind.fluent_chain > 0 then
            local call = first_kind.fluent_chain[1]
            actual_fluent = call.method .. (call.args == "" and "" or "(" .. call.args .. ")")
          end
          if expected_fluent == json.null then expected_fluent = nil end
          check_equal(actual_fluent, expected_fluent, row.id .. " bare fluent")
          for _, element in ipairs(parsed.rules[1].body) do
            check(ast.node_type(element.kind) ~= "RawBodyElementKind", row.id .. " no raw fallback")
          end
          local roundtrip = ast.from_json("SpecFile", json.decode(json.encode(ast.to_json(parsed))))
          check_same_json(ast.to_json(roundtrip), ast.to_json(parsed), row.id .. " AST JSON roundtrip")
        end

        local compiled = linkedspec.compile_spec(parsed):rule("Top")
        if expected.ownership == "action" then
          check_equal(#compiled.action_edges, #expected.targets, row.id .. " action count")
          check_equal(#compiled.blind_edges, 0, row.id .. " no blind edges")
          for index, target in ipairs(expected.targets) do
            local actual = compiled.action_edges[index].targets[1]
            check_equal(actual.label, target.label, row.id .. " action label " .. index)
            check_equal(actual.index, target.index == json.null and 0 or target.index, row.id .. " action index " .. index)
          end
        else
          check_equal(#compiled.blind_edges, #expected.targets, row.id .. " blind count")
          check_equal(#compiled.action_edges, 0, row.id .. " no action edges")
          for index, target in ipairs(expected.targets) do
            check_equal(compiled.blind_edges[index].target.label, target.label, row.id .. " blind label " .. index)
          end
        end
      end
    end
  end
end

for _, row in ipairs(contract.rule_edge_set_cases) do
  local source = edge_source(row.parent_family, array_strings(row.sources), array_strings(row.declared_rules))
  if row.expected_error ~= nil then
    local failure = validation_failure(source)
    local projected = linkedspec.spec_validation_error_to_json(failure)
    check_equal(failure.code, row.expected_error, row.id .. " code")
    check_equal(failure.stage, "validate_rule", row.id .. " stage")
    check_equal(sorted_keys(projected.fields), "ownerships,rule_label", row.id .. " fields")
    if projected.fields.ownerships ~= nil then
      check_same_json(projected.fields.ownerships, json.array({ "action", "blind" }), row.id .. " ownerships")
    end
  else
    local compiled_ok, compiled = capture(function()
      return linkedspec.compile_spec(linkedspec.parse_spec(source)):rule("Top")
    end)
    check_equal(compiled_ok, true, row.id .. " compiles: " .. tostring(compiled))
    if compiled_ok then
      if row.expected_ownership == "action" then
        check(#compiled.action_edges > 0, row.id .. " has action edges")
        check_equal(#compiled.blind_edges, 0, row.id .. " no blind edges")
      else
        check(#compiled.blind_edges > 0, row.id .. " has blind edges")
        check_equal(#compiled.action_edges, 0, row.id .. " no action edges")
      end
    end
  end
end

for _, source in ipairs({
  "Top::\n Child\n\nChild:\n /x/\n",
  "Top:: Child\n\nChild:\n /x/\n",
  "Top::AND\n Child {\n  return(child_result)\n }\n\nChild:\n /x/\n",
}) do
  local parsed = linkedspec.parse_spec(source)
  check_equal(ast.node_type(parsed.rules[1].body[1].kind), "BareEdgeBodyElementKind", "complete bare candidate")
end
local suffix = linkedspec.parse_spec("Top:: /x/ Child\n\nChild:\n /x/\n")
for _, element in ipairs(suffix.rules[1].body) do
  check(ast.node_type(element.kind) ~= "BareEdgeBodyElementKind", "bare candidate is physical-line scoped")
end

if #failures == 0 then
  io.stdout:write("rule-local cursor normalization: ", assertions, " assertions passed\n")
else
  io.stderr:write("rule-local cursor normalization: ", #failures, " of ", assertions, " assertions failed\n")
  for _, message in ipairs(failures) do io.stderr:write("- ", message, "\n") end
  os.exit(1)
end
